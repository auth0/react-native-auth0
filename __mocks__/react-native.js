/**
 * This file provides a project-wide, self-contained mock for the 'react-native' package.
 */

const React = require('react');

// A fake native module to be returned by TurboModuleRegistry.getEnforcing.
const mockNativeModule = {
  // ... (all the mock native methods from before)
  initializeAuth0WithConfiguration: jest.fn(() => Promise.resolve()),
  webAuth: jest.fn(() => Promise.resolve({})),
  webAuthLogout: jest.fn(() => Promise.resolve()),
  cancelWebAuth: jest.fn(() => Promise.resolve()),
  getBundleIdentifier: jest.fn(() => Promise.resolve('com.my.app')),
  resumeWebAuth: jest.fn(() => Promise.resolve(true)),
  saveCredentials: jest.fn(() => Promise.resolve()),
  getCredentials: jest.fn(() => Promise.resolve({})),
  hasValidCredentials: jest.fn(() => Promise.resolve(true)),
  clearCredentials: jest.fn(() => Promise.resolve()),
  hasValidInstance: jest.fn(() => Promise.resolve(true)),
};

// Export the mocked module.
module.exports = {
  // Mock the Platform module
  Platform: {
    OS: 'ios',
  },
  // Mock the TurboModuleRegistry
  TurboModuleRegistry: {
    getEnforcing: () => mockNativeModule,
  },
  // Mock the Linking module
  Linking: {
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
    getInitialURL: jest.fn(() => Promise.resolve(null)),
  },

  // FIX: Add mocks for core UI components used in tests.
  // We can use a simple functional component that just renders its children
  // and passes along any props.
  View: (props) => React.createElement('View', props, props.children),
  Text: (props) => React.createElement('Text', props, props.children),
  Button: (props) => React.createElement('Button', props, props.children),
  // Add any other components your tests might use, e.g., StyleSheet
  StyleSheet: {
    create: (styles) => styles,
  },
};
