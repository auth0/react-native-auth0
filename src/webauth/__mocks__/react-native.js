import Linking from './linking';
import A0Auth0 from './auth0';

const mock = {};

mock.Linking = new Linking();
mock.NativeModules = { A0Auth0: new A0Auth0() };
mock.Platform = { OS: 'test-os' };

module.exports = mock;
