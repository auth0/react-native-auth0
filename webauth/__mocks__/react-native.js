import Linking from './linking';
import A0Auth0 from './auth0';

const mock = {};

mock.Linking = new Linking();
mock.NativeModules = { A0Auth0: new A0Auth0() };
mock.Platform = { OS: "ios", Version: 11.1 };

module.exports = mock;