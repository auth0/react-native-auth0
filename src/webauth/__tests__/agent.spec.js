jest.mock('react-native');
import Agent from '../agent';
import {NativeModules, Linking, Platform} from 'react-native';
const A0Auth0 = NativeModules.A0Auth0;

describe('Agent', () => {
  const agent = new Agent();

  describe('show', () => {
    beforeEach(() => {
      NativeModules.A0Auth0 = A0Auth0;
      A0Auth0.reset();
    });

    it('should fail if native module is not linked', async () => {
      NativeModules.A0Auth0 = undefined;
      expect.assertions(1);
      await expect(agent.show('https://auth0.com')).rejects.toMatchSnapshot();
    });

    describe('complete web flow', () => {
      beforeEach(() => {
        A0Auth0.onUrl = () => {
          Linking.emitter.emit('url', {url: 'https://auth0.com'});
        };
      });

      it('should resolve promise with url result', async () => {
        expect.assertions(1);
        await expect(
          agent.show('https://auth0.com'),
        ).resolves.toMatchSnapshot();
      });

      it('should show initial url', async () => {
        expect.assertions(1);
        const url = 'https://auth0.com/authorize';
        await agent.show(url);
        expect(A0Auth0.url).toEqual(url);
      });

      it('should not pass ephemeral session parameter', async () => {
        expect.assertions(1);
        const url = 'https://auth0.com';
        await agent.show(url);
        expect(A0Auth0.ephemeralSession).toBeUndefined();
      });

      it('should not use ephemeral session by default', async () => {
        Platform.OS = 'ios';
        expect.assertions(1);
        const url = 'https://auth0.com';
        await agent.show(url);
        expect(A0Auth0.ephemeralSession).toEqual(false);
      });

      it('should set ephemeral session', async () => {
        Platform.OS = 'ios';
        expect.assertions(1);
        const url = 'https://auth0.com';
        await agent.show(url, true);
        expect(A0Auth0.ephemeralSession).toEqual(true);
      });
    });

    describe('listeners', () => {
      it('should register url listeners', () => {
        A0Auth0.onUrl = () => {};
        const url = 'https://auth0.com/authorize';
        agent.show(url);
        expect(Linking.emitter.listenerCount('url')).toEqual(1);
      });

      it('should remove url listeners when done', async () => {
        A0Auth0.onUrl = () => {
          Linking.emitter.emit('url', {url: 'https://auth0.com'});
        };
        expect.assertions(1);
        const url = 'https://auth0.com/authorize';
        await agent.show(url);
        expect(Linking.emitter.listenerCount('url')).toEqual(0);
      });

      it('should remove url listeners when load fails', async () => {
        expect.assertions(1);
        A0Auth0.error = new Error('failed to load');
        const url = 'https://auth0.com/authorize';
        await agent.show(url).catch(err => Promise.resolve(err));
        expect(Linking.emitter.listenerCount('url')).toEqual(0);
      });

      it('should remove url listeners on first load', async () => {
        expect.assertions(1);
        const url = 'https://auth0.com/authorize';
        await agent.show(url, false, true);
        expect(Linking.emitter.listenerCount('url')).toEqual(0);
      });
    });

    describe('failed flow', () => {
      it('should reject with error', async () => {
        expect.assertions(1);
        A0Auth0.error = new Error('failed to load');
        await expect(agent.show('https://auth0.com')).rejects.toMatchSnapshot();
      });
    });
  });

  describe('newTransaction', () => {
    it('should call native integration', async () => {
      const parameters = {state: 'state'};
      A0Auth0.parameters = parameters;
      expect.assertions(1);
      await expect(agent.newTransaction()).resolves.toMatchSnapshot();
    });

    it('should fail if native module is not linked', async () => {
      NativeModules.A0Auth0 = undefined;
      expect.assertions(1);
      await expect(agent.newTransaction()).rejects.toMatchSnapshot();
    });
  });
});
