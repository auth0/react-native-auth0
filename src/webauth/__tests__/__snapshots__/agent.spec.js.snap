// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Agent login should fail if native module is not linked 1`] = `[Error: Missing NativeModule. React Native versions 0.60 and up perform auto-linking. Please see https://github.com/react-native-community/cli/blob/master/docs/autolinking.md.]`;

exports[`Agent logout should fail if native module is not linked 1`] = `[Error: Missing NativeModule. React Native versions 0.60 and up perform auto-linking. Please see https://github.com/react-native-community/cli/blob/master/docs/autolinking.md.]`;